File: leetcode/Maximum Depth of Binary Tree.js
// Given a binary tree, find its maximum depth.
//
// The maximum depth is the number of nodes along the longest path from the root node down to the farthest leaf node.
//


/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function(root) {
    if (!root) return 0;

    function recursive(root, depth) {
        if (!root.left && !root.right) {
            return depth;
        }
        let result_left = 0,
            result_right = 0;
        if (root.left) {
            result_left = recursive(root.left, depth + 1);
        }
        if (root.right) {
            result_right = recursive(root.right, depth + 1);
        }
        return Math.max(result_left, result_right);
    }

    return recursive(root, 1);
};


var maxDepth_BFS = function(root) {
    if (!root) {
        return 0;
    }
    let queue = [];
    queue.push(root);
    let count = 0;
    while (queue.length) {
        let size = queue.length;
        while (size-- > 0) {
            let node = queue.shift();
            if (node.left) {
                queue.push(node.left);
            }
            if (node.right) {
                queue.push(node.right);
            }
        }
        count++;
    }
    return count;
};

var maxDepth_DFS = function(root) {
    if(!root) {
        return 0;
    }

    let stack = [];
    let value = [];
    stack.push(root);
    value.push(1);
    let max = 0;
    while(stack.length) {
        let node = stack.pop();
        let temp = value.pop();
        max = Math.max(temp, max);
        if(node.left) {
            stack.push(node.left);
            value.push(temp+1);
        }
        if(node.right) {
            stack.push(node.right);
            value.push(temp+1);
        }
    }
    return max;
};

File: leetcode/Maximum Depth of Binary Tree.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# LC104
# https://leetcode.com/problems/maximum-depth-of-binary-tree/description/


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        
        def recursive(root, depth):
            if not root.left and not root.right:
                return depth
            
            result_left = 0
            result_right = 0
            
            if root.left:
                result_left = recursive(root.left, depth + 1)
            if root.right:
                result_right = recursive(root.right, depth + 1)
            return max(result_left, result_right)
        
        return recursive(root, 1)
    
    def maxDepth2(self, root):
        if not root:
            return 0
        
        queue = [root]
        count = 0
        while len(queue) > 0:
            size = len(queue)
            while size > 0:
                size -= 1
                node = queue.pop(0)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            
            count += 1
        
        return count
    
    def maxDepth3(self, root):
        if not root:
            return 0
        
        stack = [root]
        value = [1]
        
        max_val = 0
        
        while len(stack):
            node = stack.pop()
            temp = value.pop()
            
            max_val = max(temp, max_val)
            
            if node.left:
                stack.append(node.left)
                value.append(temp + 1)
            
            if node.right:
                stack.append(node.right)
                value.append(temp + 1)
        
        return max_val
